docs/design/2021-12-08-instance-scope.md

```diff
@@ -0,0 +1,224 @@
+
+### Stage 3: Refactoring
+
+The use of a `GetGlobal()` and `SetGlobal()` func for each instance scoped system variable is not ideal. It is possible to refactor the system variable framework so that instance scope is stored in a map, and the values are updated automatically by `SET GLOBAL` on an instance scoped variable. On startup, as the configuration file is parsed it will update the values in the map. This seems like a better approach than the current use of Setters/Getters, and because there is a prescribed way of doing it we can correctly handle the data races that are common with our current incorrect usage of calling `config.GetGlobalConfig()`.
+
+Thus, the source of truth for instance scoped variables moves from the `config` package to another part of the server (likely `domain`). See [issue #30366](https://github.com/pingcap/tidb/issues/30366).
+
+## Impacts & Risks
+
+The biggest risk is that we can not agree on a reduced scope of implementation and the project extends to cover increased scope. Configuration management is a classic example of a [bikeshed problem](https://en.wikipedia.org/wiki/Law_of_triviality), and we will likely need to make some tradeoffs to get anywhere.
+
+Many of the alternatives are difficult to implement because they break compatibility (we need to live with `GLOBAL` means cluster `GLOBAL`, and there is no `SET CLUSTER`) or they break rolling upgrade scenarios because we currently have no specific rules of the minimum version of TiDB which can be upgraded from.
+
+## Investigation & Alternatives
+
+- There is no equivalent functionality to reference in MySQL since it does not have the native concept of a cluster.
+- An [earlier proposal](https://docs.google.com/document/d/1RuajYAFVsjJCwCBpIYF--9jtgxDZZcz3cbP-iVuLxJI/edit) for `INSTANCE` scope, with rules for precedence (same author)
+- CockroachDB has cluster level settings and node-level settings. Most settings are cluster level, and [node-level](https://www.cockroachlabs.com/docs/v21.2/cockroach-start) needs to be parsed as arguments when starting the server.
+
+## Unresolved Questions
+
+There are **very few** known scenarios where a system variable is *currently both* instance and global scoped, but these individually would need to be handled:
+
+```
+package main
+
+import (
+	"fmt"
+
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/pingcap/tidb/sessionctx/variable"
+)
+
+func main() {
+	for _, v := range variable.GetSysVars() {
+		if v.HasGlobalScope() && v.HasSessionScope() && (v.GetSession != nil || v.SetSession != nil) && (v.GetGlobal != nil || v.SetGlobal != nil) {
+			fmt.Printf("%s\n", v.Name)
+		}
+	}
+}
+```
+
+**Output:**
+
+```
+tidb_store_limit
+tidb_stmt_summary_internal_query
+tidb_enable_stmt_summary
+tidb_stmt_summary_max_sql_length
+tidb_stmt_summary_max_stmt_count
+tidb_capture_plan_baselines
+tidb_stmt_summary_refresh_interval
+tidb_stmt_summary_history_size
+```
+
+The following suggestions are provided (to be discussed):
```

Comment:
There are many more, such as:
- the system variables controlling general log, slow log, top SQL
- `tidb_pprof_sql_cpu`, `tidb_memory_usage_alarm_ratio`.
- `tidb_tso_client_batch_max_wait_time`, `tidb_enable_tso_follower_proxy`, `tidb_enable_local_txn`.
- other system variables controlling statement summary except for `tidb_enable_stmt_summary`, such as `tidb_stmt_summary_internal_query`, `tidb_stmt_summary_refresh_interval`, `tidb_stmt_summary_history_size`, `tidb_stmt_summary_max_stmt_count`, `tidb_stmt_summary_max_sql_length`.